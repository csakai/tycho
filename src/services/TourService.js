import Cookies from 'js-cookie';

export default class TourService {

  /**
   * Determines if the user has opted, via cookie, to skip the tour.
   *
   * @returns {Boolean} ability to skip tour
   */
  static canSkip = () => {
    return Cookies.get('tourViewed') === 'true';
  }

  /**
   * Sets a cookie to indicate that the tour was either viewed or skipped.
   */
  static setSkip = () => {
    Cookies.set('tourViewed', 'true', {expires: 365});
  }

  /**
   * Calculates the total time that showing labels will take,
   * separated and padded by the SEPARATION_INTERVAL constant.
   *
   * @param {Object[]} labels - list of labels to sum
   * @returns {Number} total duration
   */
  static getTourDuration = (labels) => {
    const SEPARATION_INTERVAL = 1000;

    let duration = SEPARATION_INTERVAL;
    
    duration += labels.reduce((cur, next) => {
      return cur + next.duration + SEPARATION_INTERVAL;
    }, 0);
    duration += SEPARATION_INTERVAL;

    return duration;
  }
}
