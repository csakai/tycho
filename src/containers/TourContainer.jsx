import React from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import * as UIControlsActions from '../actions/UIControlsActions';
import * as TourActions from '../actions/TourActions';
import * as LabelActions from '../actions/LabelActions';
import ReduxService from '../services/ReduxService';
import TourService from '../services/TourService';
import TourLabelContainer from './TourLabelContainer';
import Tour from '../components/Tour';
import Constants from '../constants';

export class TourContainer extends React.Component {

  static propTypes = {
      labels: PropTypes.array
  }

  componentDidMount = () => {
      if (TourService.canSkip()) {
          this.props.action.tourSkipped(true);
      }
  }

  componentWillReceiveProps = (nextProps) => {
      this.maybeSkipTour(nextProps);
      this.maybeStartTour(nextProps);
  }

  /**
   * Skips tour if opted to do so and is not already skipped.
   * 
   * @param {Object} nextProps
   * @param {Boolean} nextProps.isSkipped
   */
  maybeSkipTour = ({isSkipped}) => {
      if (this.props.isSkipped !== isSkipped && isSkipped) {
          this.skipTour();
      }
  }

  /**
   * Starts tour if not already initialized.
   * 
   * @param {Object} nextProps
   * @param {Boolean} nextProps.playing
   */
  maybeStartTour = ({playing}) => {
      if (playing && !this.props.playing && !this.props.isComplete) {
          this.initializeTour();
      }
  }

  /**
   * Returns true if the tour should be run.
   * 
   * @returns {Boolean}
   */
  shouldRunTour = () => {
      return this.props.playing && !this.props.isSkipped;
  }

  /**
   * Initializes the tour.
   */
  initializeTour = () => {
      const {action, labels} = this.props;
      const tourDuration = TourService.getTourDuration(labels);

      if (!TourService.canSkip()) {
          action.setUIControls(false);
          action.setCameraOrbit(true);

          setTimeout(this.onOrbitComplete, tourDuration);
      }
  }
  
  /**
   * Callback to invoke once the tour orbit has completed.
   */
  onOrbitComplete = () => {
      if (!this.props.isComplete) {
          this.setDefaultActiveOrbital();
          setTimeout(this.onTourComplete, Constants.WebGL.Tween.SLOW);
      }
  }

  /**
   * Callback to invoke once the entire tour is completed.
   */
  onTourComplete = () => {
      this.props.action.tourCompleted(true);
  }

  /**
   * Skips the tour and zooms directly into the target planet.
   */
  skipTour = () => {
      const {action} = this.props;

      action.tourCompleted(true);
      action.setCameraOrbit(false);
      action.setUIControls(true);

      this.setDefaultActiveOrbital();
  }

  /**
   * Sets the active orbital targetId and the header label text to UI defaults.
   */
  setDefaultActiveOrbital = () => {
      this.props.action.setActiveOrbital(Constants.UI.ALTERNATE_TARGET_NAME);
      this.props.action.setActiveOrbital(Constants.UI.DEFAULT_TARGET_NAME);
      this.props.action.setLabelText(Constants.UI.DEFAULT_LABEL_TEXT);
  }

  /**
   * Sets the skip tour cookie and skips tour.
   */
  skipTourTrigger = () => {
      TourService.setSkip();
      this.skipTour();
  }
  
  /**
   * Maps the labels list prop to TourLabelContainers.
   *
   * @param {Object[]} labels - list of labels
   * @returns {TourLabelContainer[]}
   */
  getLabels = (labels) => {
      const separation = Constants.Tour.SEPARATION_INTERVAL;
      let totalTime = separation;
    
      return labels.map(({text, duration}, key) => {
          totalTime += separation;
          const start = totalTime;
          totalTime += duration;
          const end = totalTime;

          return (
              <TourLabelContainer
                  key={key}
                  text={text}
                  start={start}
                  end={end}
              />
          );
      });
  }

  render() {
      if (!this.shouldRunTour()) {
          return null;
      }
      return (
          <Tour
              {...this.props}
              skipTour={this.skipTourTrigger}
              labels={this.getLabels(this.props.labels)}
          />
      );
  }
}

export default connect(
    ReduxService.mapStateToProps(
        'uiControls.controlsEnabled',
        'uiControls.scale',
        'label.targetId',
        'tour.isComplete',
        'tour.isSkipped',
        'animation.playing',
        'data.pageText'
    ),
    ReduxService.mapDispatchToProps(
        UIControlsActions,
        TourActions,
        LabelActions
    )
)(TourContainer);