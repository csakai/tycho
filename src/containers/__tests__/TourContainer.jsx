import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import {TourContainer} from '../TourContainer';
import TourService from '../../services/TourService';

const labels = [
  {
    duration: 5000,
    text: 'Welcome to the Solar System'
  },
  {
    duration: 5000,
    text: 'This is a real-time interactive simulation of major planetary bodies'
  },
  {
    duration: 3000,
    text: 'Let\'s start exploring'
  }
];

jest.useFakeTimers();

describe('Tour Container', () => {
  let component, tourContainer, action;

  beforeEach(() => {
    action = {
      setUIControls: jest.fn(),
      setCameraOrbit: jest.fn(),
      setActiveOrbital: jest.fn(),
      tourCompleted: jest.fn(),
      tourSkipped: jest.fn(),
      showLabel: jest.fn()
    };

    component = shallow(<TourContainer
      labels={labels}
      action={action}
    />);
    tourContainer = component.instance();
  });

  describe('componentDidMount()', () => {
    it('should call the tour skip action if the tour can be skipped', () => {
      TourService.canSkip = () => true;
      const spy = jest.spyOn(tourContainer.props.action, 'tourSkipped');

      tourContainer.componentDidMount();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(true);;
    });

    it('should initialize the tour if the tour cannot be skipped', () => {
      TourService.canSkip = () => false;
      const spy = jest.spyOn(tourContainer, 'initializeTour');

      tourContainer.componentDidMount();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
    });

  });

  describe('initializeTour()', () => {
    beforeEach(() => {
      tourContainer.getTourDuration = jest.fn();
    });

    it('should call setUIControls with false', () => {
      const spy = jest.spyOn(tourContainer.props.action, 'setUIControls');

      tourContainer.initializeTour();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(false);
    });

    it('should call setCameraOrbit with true', () => {
      const spy = jest.spyOn(tourContainer.props.action, 'setCameraOrbit');

      tourContainer.initializeTour();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(true);
    });

    it('should call onOrbitComplete after the calculated duration has passed', () => {
      tourContainer.onOrbitComplete = jest.fn();
      const spy = jest.spyOn(tourContainer, 'onOrbitComplete');
      const duration = 5000;

      tourContainer.initializeTour();

      expect(spy).not.toHaveBeenCalled();

      jest.runAllTimers();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });

  describe('componentWillReceiveProps()', () => {
    let spy;

    beforeEach(() => {
      spy = jest.spyOn(tourContainer, 'skipTour');
    });

    it('should call skipTour() if the isSkipped prop has changed and is true', () => {
      tourContainer.componentWillReceiveProps({isSkipped: true});

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should not call skipTour() if the isSkipped prop is false', () => {
      tourContainer.componentWillReceiveProps({isSkipped: false});
      expect(spy).not.toHaveBeenCalled();
    });

    it('should not call skipTour() if the isSkipped prop has not changed', () => {
      tourContainer.componentWillReceiveProps({});
      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('onOrbitComplete()', () => {
    it('should call setActiveOrbital with the default orbital name', () => {
      const spy = jest.spyOn(tourContainer.props.action, 'setActiveOrbital');

      tourContainer.onOrbitComplete();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith('dummyParent'); // TODO: const
    });
  });

  describe('onTourComplete()', () => {
    it('should call tourCompleted with true', () => {
      const spy = jest.spyOn(tourContainer.props.action, 'tourCompleted');

      tourContainer.onTourComplete();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(true);
    });
  });

  describe('skipTour()', () => {
    it('should call the tourCompleted action with `true`', () => {
      const spy = jest.spyOn(tourContainer.props.action, 'tourCompleted');

      tourContainer.skipTour();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(true);
    });

    it('should call the setCameraOrbit action with `false`', () => {
      const spy = jest.spyOn(tourContainer.props.action, 'setCameraOrbit');

      tourContainer.skipTour();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(false);
    });

    it('should call the setActiveOrbital action with the defined targetName', () => {
      const spy = jest.spyOn(tourContainer.props.action, 'setActiveOrbital');

      tourContainer.skipTour();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith('dummyParent'); // TODO
    });
  });

  describe('skipTourTrigger()', () => {
    it('should call skipTour()', () => {
      const spy = jest.spyOn(tourContainer, 'skipTour');

      tourContainer.skipTourTrigger();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });

  describe('getLabels()', () => {
    it('should return an array of TourLabels', () => {
      const result = tourContainer.getLabels(labels);

      expect(Array.isArray(result)).toBe(true);

      result.forEach((label) => {
        expect(label).toMatchSnapshot();
      });
    });
  });

  describe('getModifier()', () => {
    it('should return `skip` if isSkipped = true', () => {
      expect(tourContainer.getModifier({
        isSkipped: true
      })).toEqual('skip');
    });

    it('should return `show` if isComplete = false', () => {
      expect(tourContainer.getModifier({
        isComplete: false
      })).toEqual('show');
    });

    it('should return `hide` if isComplete = true', () => {
      expect(tourContainer.getModifier({
        isComplete: true
      })).toEqual('hide');
    });
  });

  describe('render()', () => {
    it('should render the tour container successfully', () => {
      component.setState({time: 1});
      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
