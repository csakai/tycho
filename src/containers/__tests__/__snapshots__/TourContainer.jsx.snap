// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tour Container getLabels() should return an array of TourLabels 1`] = `
<TourLabelContainer
  end={6000}
  start={1000}
  text="Welcome to the Solar System"
/>
`;

exports[`Tour Container getLabels() should return an array of TourLabels 2`] = `
<TourLabelContainer
  end={12000}
  start={7000}
  text="This is a real-time interactive simulation of major planetary bodies"
/>
`;

exports[`Tour Container getLabels() should return an array of TourLabels 3`] = `
<TourLabelContainer
  end={16000}
  start={13000}
  text="Let's start exploring"
/>
`;

exports[`Tour Container render() should render the tour container with the hide modifier if isComplete = true 1`] = `
<div>
  <div
    className="tour tour--hide"
  >
    <div
      className="tour__theater-bar tour__theater-bar--hide"
      style={
        Object {
          "top": 0,
        }
      }
    />
    <div
      className="tour__labels"
    >
      <TourLabelContainer
        end={6000}
        key="0"
        start={1000}
        text="Welcome to the Solar System"
      />
      <TourLabelContainer
        end={12000}
        key="1"
        start={7000}
        text="This is a real-time interactive simulation of major planetary bodies"
      />
      <TourLabelContainer
        end={16000}
        key="2"
        start={13000}
        text="Let's start exploring"
      />
    </div>
    <div
      className="tour__theater-bar tour__theater-bar--hide"
      style={
        Object {
          "bottom": 0,
        }
      }
    >
      <span
        className="tour__skip-link"
        onClick={[Function]}
      >
        Skip Tour
      </span>
    </div>
  </div>
  <Connect(SpinLabelContainer) />
</div>
`;

exports[`Tour Container render() should render the tour container with the show modifier if isComplete = false 1`] = `
<div>
  <div
    className="tour tour--show"
  >
    <div
      className="tour__theater-bar tour__theater-bar--show"
      style={
        Object {
          "top": 0,
        }
      }
    />
    <div
      className="tour__labels"
    >
      <TourLabelContainer
        end={6000}
        key="0"
        start={1000}
        text="Welcome to the Solar System"
      />
      <TourLabelContainer
        end={12000}
        key="1"
        start={7000}
        text="This is a real-time interactive simulation of major planetary bodies"
      />
      <TourLabelContainer
        end={16000}
        key="2"
        start={13000}
        text="Let's start exploring"
      />
    </div>
    <div
      className="tour__theater-bar tour__theater-bar--show"
      style={
        Object {
          "bottom": 0,
        }
      }
    >
      <span
        className="tour__skip-link"
        onClick={[Function]}
      >
        Skip Tour
      </span>
    </div>
  </div>
  <Connect(SpinLabelContainer) />
</div>
`;
